File: python/src/st_device.py
#
# STAR TREK Text Game
# 2017-05-01 K.OHWADA
#

import random	

## class
class Device():


	
	DEVICES = [ "L.R. SENSOR",   "S.R. SENSOR", "PHASER CNTRL", "TORPEDO TUBES", "SHIELD CNTRL", "DAMAGE CNTRL", "WARP ENGINE", "IMPULSE ENGINE", "COMPUTER" ]
	 
	SIZE_D = 9
	
	# 30 %
	DAMEGE_PROBABILITY = 0.3
	
# very easy : 50 %
	SHORT_SENSOR_REPAIR_PROBABILITY = 0.5
		
	DEVICE_LONG_SENSOR = 0
	DEVICE_SHORT_SENSOR = 1
	DEVICE_PHASER = 2
	DEVICE_TORPEDO = 3
	DEVICE_SHIELD = 4
	DEVICE_DAMAGE = 5
	DEVICE_WARP = 6
	DEVICE_IMPULSE = 7
	DEVICE_COMPUTER = 8 

# damege level : 0 , pluse: normal  minus: dameged
	d_arr = []
	
	def __init__(self):
		pass				
### def end

	def setup(self):
		self.d_arr = [0 for i in range(self.SIZE_D)]	
### def end

	def print_damege_report(self):
		print
		print "DAMAGE CONTROL REPORT"
		for i in range(self.SIZE_D):
			status = "Normal" if self.is_available(i) else "Dameged"
			msg = self.DEVICES[i] + " : " + status
			print msg
### for i
### def end

	def is_available(self, id):
		ret = True if ( self.d_arr[id] >= 0 ) else False
		return ret
### def end

	def check_available(self, id):
		ret = False
		if ( self.is_available(id) ) :
			ret = True
		else:
			msg = self.DEVICES[id] + " dameged"
			print msg
### if end			
		return ret
### def end	

	def change_available(self):
		
		### 0 - 8
		id = int((self.SIZE_D -1)*random.random() )
		
	### 1 - 6
		level = int(5*random.random() ) + 1
		
		if ( random.random() <  self.DAMEGE_PROBABILITY ):
			self.d_arr[id] -= level
		else:
			self.d_arr[id] += level
			
# repair specially
# because player can do nothing
# when SR Sensor breaks
		if ( random.random() <  	self.SHORT_SENSOR_REPAIR_PROBABILITY ):
			self.d_arr[	self.DEVICE_SHORT_SENSOR ] = 0			
### def end			


	def damege_all(self):	
		self.set_available_all( -1)	
### def end	

	def repair_all(self):	
		self.set_available_all( 0 )	
### def end

	def set_available_all(self, value):	
			for i in range(self.SIZE_D):
				self.d_arr[i] = value
### def end
				
### class end

File: javascript/src/js/st_device.js
//
// STAR TREK Text Game
// 2017-06-01 K.OHWADA
//

var Device = function () {

	this.DEVICES = [ "L.R. SENSOR",   "S.R. SENSOR", "PHASER CNTRL", "TORPEDO TUBES", "SHIELD CNTRL", "DAMAGE CNTRL", "WARP ENGINE", "IMPULSE ENGINE", "COMPUTER" ];
	
	this.DEBUG_AVAILABLE = true;
		
	this.DEVICE_LONG_SENSOR = 0;
	this.DEVICE_SHORT_SENSOR = 1;
	this.DEVICE_PHASER = 2;
	this.DEVICE_TORPEDO = 3;
	this.DEVICE_SHIELD = 4;
	this.DEVICE_DAMAGE = 5;
	this.DEVICE_WARP = 6;
	this.DEVICE_IMPULSE = 7;
	this.DEVICE_COMPUTER = 8;

	// damege level : 0 , pluse: normal  minus: dameged
	this.d_arr = [];
	
	} // constractor	

		
	Device.prototype.setup = function () {
		var k =0;
		for(k=0;k < 9;k++) {
		this.d_arr[k] = 0;	
	} // for k
	
	} // setup 		


	Device.prototype.damege_report = function () {

		
		var msg = "DAMAGE CONTROL REPORT<br/>\n";
		
				var k =0;
				var status = "";
				
		for(k=0;k < 9;k++) {
			status = (this.is_available(k) ) ? "Normal" : "Dameged";
			msg += this.DEVICES[k] + " : " + status + "<br/>\n" ;
	} // for k

	return msg;
} // damege_report


Device.prototype.is_available = function (id) {
		if ( this.DEBUG_AVAILABLE ) {
			return true;
		} // if
		
		var ret = ( this.d_arr[id] >= 0 ) ? true :  false ;
		return ret;
		
} // is_available		


Device.prototype.check_available = function (id) {

		var ret = false;
		var msg = "";
		if ( this.is_available(id) ) {
			ret = true;
		} else {
			msg = this.DEVICES[id] + " dameged";
			alert( msg );
		
		} // if
					
		return ret ;

} // check_available


	
	Device.prototype.change_available = function () {
		
		// 0 - 8
		var id = Math.floor(8*Math.random() ) ;
		
	// 1 - 6
		var level = Math.floor(5*Math.random() ) + 1 ;
		
		if ( Math.random() <  0.3 ) {
			this.d_arr[id] -= level ;
		} else {
			this.d_arr[id] += level ;
		} // if
			
// repair specially
// because player can do nothing
// when SR Sensor breaks
		if ( Math.random() < 0.3  ) {
			this.d_arr[	this.DEVICE_SHORT_SENSOR ] = 0	;
		} // if
						
} // change_available	
	

	Device.prototype.damege_all = function () {
		this.set_available_all( -1);	
} // damege_all


	Device.prototype.repair_all = function () {	
		this.set_available_all( 0 );	
} // repair_all 


	Device.prototype.set_available_all = function (value) {	

				var k =0;
				
		for(k=0;k < 9;k++){
					this.d_arr[k] = value;
		} // k
								
} // set_available_all



Device.prototype.damege_phaser = function () {	
	
	// 50    5
	if ( Math.random() < 0.5 ) {
this.d_arr[ this.DEVICE_PHASER ] = -1;
 }// if 
								
} // damege_phaser